File: app/src/main/java/com/example/eziketobenna/bakingapp/ui/recipes/RecipeAdapter.java
package com.example.eziketobenna.bakingapp.ui.recipes;

import androidx.databinding.DataBindingUtil;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.example.eziketobenna.bakingapp.R;
import com.example.eziketobenna.bakingapp.data.model.Recipe;
import com.example.eziketobenna.bakingapp.databinding.ContentMainBinding;

import java.util.List;

public class RecipeAdapter extends RecyclerView.Adapter<RecipeAdapter.ViewHolder> {
    private List<Recipe> mRecipeList;
    private RecipeClickListener mListener;

    RecipeAdapter(RecipeClickListener listener) {
        mListener = listener;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        LayoutInflater layoutInflater = LayoutInflater.from(viewGroup.getContext());
        ContentMainBinding binding = DataBindingUtil.inflate(layoutInflater, R.layout.content_main, viewGroup,
                false);
        return new ViewHolder(binding.getRoot());
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int i) {
        Recipe recipe = mRecipeList.get(i);
        holder.bind(recipe);
    }

    @Override
    public int getItemCount() {
        return mRecipeList == null ? 0 : mRecipeList.size();
    }


    public void setRecipes(List<Recipe> recipes) {
        mRecipeList = recipes;
        notifyDataSetChanged();
    }

    public interface RecipeClickListener {
        void onRecipeClick(Recipe recipe);
    }

    class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        private final ContentMainBinding binding;

        ViewHolder(@NonNull View itemView) {
            super(itemView);
            binding = DataBindingUtil.bind(itemView);
            itemView.setOnClickListener(this);
        }

        void bind(Recipe recipe) {
            binding.setRecipe(recipe);
            binding.executePendingBindings();
        }

        @Override
        public void onClick(View v) {
            int position = getAdapterPosition();
            Recipe recipe = mRecipeList.get(position);
            mListener.onRecipeClick(recipe);
        }
    }
}

File: features/recipes/recipe/src/main/java/com/example/eziketobenna/bakingapp/recipe/ui/RecipeAdapter.kt
package com.example.eziketobenna.bakingapp.recipe.ui

import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.example.eziketobenna.bakingapp.core.imageLoader.ImageLoader
import com.example.eziketobenna.bakingapp.recipe.R
import com.example.eziketobenna.bakingapp.recipe.databinding.ContentMainBinding
import com.example.eziketobenna.bakingapp.recipe.inflate
import com.example.eziketobenna.bakingapp.recipe.notEmpty
import com.example.eziketobenna.bakkingapp.model.model.RecipeModel
import javax.inject.Inject

typealias RecipeClickListener = (RecipeModel) -> Unit

class RecipeAdapter @Inject constructor(private val imageLoader: ImageLoader) :
    ListAdapter<RecipeModel, RecipeAdapter.RecipeViewHolder>(diffUtilCallback) {

    var clickListener: RecipeClickListener? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecipeViewHolder {
        return RecipeViewHolder(ContentMainBinding.bind(parent.inflate(R.layout.content_main)))
    }

    override fun onBindViewHolder(holder: RecipeViewHolder, position: Int) {
        holder.bind(getItem(position), imageLoader, clickListener)
    }

    class RecipeViewHolder(private val binding: ContentMainBinding) :
        RecyclerView.ViewHolder(binding.root) {

        fun bind(
            model: RecipeModel,
            imageLoader: ImageLoader,
            clickListener: RecipeClickListener?
        ) {
            binding.supportingText.text = model.name
            val imageUrl: String = getImageURL(model)
            imageUrl.notEmpty { url ->
                imageLoader.loadImage(binding.recipeImage, url)
            }
            binding.root.setOnClickListener {
                clickListener?.invoke(model)
            }
        }

        private fun getImageURL(model: RecipeModel): String {
            val imageURLs: Array<String> =
                binding.root.context.resources.getStringArray(R.array.images)
            return when {
                model.image.isNotEmpty() -> model.image
                adapterPosition <= imageURLs.lastIndex -> imageURLs[adapterPosition]
                else -> ""
            }
        }
    }

    companion object {
        val diffUtilCallback: DiffUtil.ItemCallback<RecipeModel>
            get() = object : DiffUtil.ItemCallback<RecipeModel>() {
                override fun areItemsTheSame(oldItem: RecipeModel, newItem: RecipeModel): Boolean {
                    return oldItem === newItem && oldItem.id == newItem.id
                }

                override fun areContentsTheSame(
                    oldItem: RecipeModel,
                    newItem: RecipeModel
                ): Boolean {
                    return oldItem == newItem
                }
            }
    }
}
